titus-server-master/src/main/java/com/netflix/titus/master/mesos/kubeapiserver/direct/DefaultKubeApiFactory.java

```diff
@@ -0,0 +1,180 @@
+/*
+ * Copyright 2020 Netflix, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.titus.master.mesos.kubeapiserver.direct;
+
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+import javax.inject.Singleton;
+
+import com.netflix.titus.common.util.ExceptionExt;
+import com.netflix.titus.common.util.guice.annotation.Deactivator;
+import io.kubernetes.client.ApiClient;
+import io.kubernetes.client.apis.CoreV1Api;
+import io.kubernetes.client.informer.SharedIndexInformer;
+import io.kubernetes.client.informer.SharedInformerFactory;
+import io.kubernetes.client.models.V1Node;
+import io.kubernetes.client.models.V1NodeList;
+import io.kubernetes.client.models.V1Pod;
+import io.kubernetes.client.models.V1PodList;
+import io.kubernetes.client.util.CallGeneratorParams;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static com.netflix.titus.master.mesos.kubeapiserver.KubeUtil.createSharedInformerFactory;
+
+@Singleton
+public class DefaultKubeApiFactory implements KubeApiFactory {
```

Comment:
minor nit: first time I read the code, the name `Factory` implied to me that new instances would be returned for each call, since this is providing singletons (memoized instances), maybe rename this class to `...Provider`, or `...Holder`?

Reply:
`Provider` would be better suffix, but it is used by Guice, and it would be confusing.
I will try to find a better name.

Reply:
well, I already used `...Provider` in some other places :-)